packages/perseus/src/widgets/interactive-graphs/reducer/interactive-graph-reducer.ts

```diff
@@ -0,0 +1,244 @@
+import {vector as kvector} from "@khanacademy/kmath";
+import {UnreachableCaseError} from "@khanacademy/wonder-stuff-core";
+import {vec} from "mafs";
+
+import {
+    MOVE_CONTROL_POINT,
+    type InteractiveGraphAction,
+    MOVE_ALL,
+    MOVE_LINE,
+} from "./interactive-graph-action";
+
+import type {CollinearTuple} from "../../../perseus-types";
+import type {InteractiveGraphState} from "../types";
+import type {Interval} from "mafs";
+
+/** Determine if coords is type CollinearTuple[] */
+const isCollinearTuples = (
+    coords: readonly CollinearTuple[] | readonly vec.Vector2[],
+): coords is readonly CollinearTuple[] => Array.isArray(coords[0][0]);
+
+// Generic type makes returned state match input state
+export function interactiveGraphReducer<
+    GraphState extends InteractiveGraphState,
+>(state: Readonly<GraphState>, action: InteractiveGraphAction): GraphState {
+    const {snapStep, range} = state;
+    if (!state.coords) {
+        throw new Error("Graph state must have been initialized with coords");
+    }
+    switch (action.type) {
+        case MOVE_CONTROL_POINT: {
+            const newCoords =
+                action.objectIndex !== undefined &&
+                state.coords &&
+                isCollinearTuples(state.coords)
+                    ? updateAtIndex({
+                          array: state.coords,
+                          index: action.objectIndex,
+                          update: (tuple) =>
+                              setAtIndex({
+                                  array: tuple,
+                                  index: action.pointIndex,
+                                  newValue: snap({
+                                      snapStep,
+                                      point: bound({
+                                          snapStep,
+                                          range,
+                                          point: action.destination,
+                                      }),
+                                  }),
+                              }),
+                      })
+                    : setAtIndex({
+                          array: state.coords,
+                          index: action.pointIndex,
+                          newValue: snap({
+                              snapStep,
+                              point: bound({
+                                  snapStep,
+                                  range,
+                                  point: action.destination,
+                              }),
+                          }),
+                      });
+            if (
+                coordsOverlap(
+                    isCollinearTuples(newCoords) ? newCoords.flat() : newCoords,
+                )
```

Comment:
I see, thanks for the context. I thought about that, but wagered we wouldn't have questions like that since it makes the bottom one inaccessible. I'll change this.